File: chat-client/src/chat_client/messages.py
from __future__ import annotations

from dataclasses import asdict, dataclass, is_dataclass
from typing import Any, Dict, Generic, Literal, TypeVar


PayloadT = TypeVar("PayloadT")


class WebSocketEnvelope(Generic[PayloadT]):
    """기본 웹소켓 메시지 포맷 (type + payload)을 표현."""

    #: 자식 클래스에서 literals로 덮어쓰는 것을 기대한다.
    type: str
    payload: PayloadT

    def __init__(self, *, type: str, payload: PayloadT) -> None:
        self.type = type
        self.payload = payload

    def to_dict(self) -> Dict[str, Any]:
        payload = self.payload
        if is_dataclass(payload):
            payload = asdict(payload)
        return {"type": self.type, "payload": payload}


class WebSocketMessageTypes:
    MESSAGE_SEND: Literal["MESSAGE_SEND"] = "MESSAGE_SEND"
    MESSAGE_RECEIVE: Literal["MESSAGE_RECEIVE"] = "MESSAGE_RECEIVE"
    SERVER_STATUS_REQUEST: Literal["SERVERSTATUS_REQUEST"] = "SERVERSTATUS_REQUEST"
    SERVER_STATUS_RESPONSE: Literal["SERVERSTATUS_RESPONSE"] = "SERVERSTATUS_RESPONSE"
    SYSTEM_MESSAGE_RECEIVE: Literal["SYSTEM_MESSAGE_RECEIVE"] = "SYSTEM_MESSAGE_RECEIVE"


@dataclass(slots=True)
class MessageSendPayload:
    """사용자가 서버에 보내는 채팅 메시지."""

    message: str


@dataclass(slots=True)
class MessageReceivePayload:
    """서버가 브로드캐스트하는 채팅 메시지."""

    message: str
    nickname: str
    timestamp: str


@dataclass(slots=True)
class ServerStatusRequestPayload:
    """서버 인원 수 조회 요청 (내용 없음)."""

    pass


@dataclass(slots=True)
class ServerStatusResponsePayload:
    """현재 서버 접속자 수 응답."""

    clientCount: int


@dataclass(slots=True)
class SystemMessageReceivePayload:
    """시스템에서 내려오는 공지 메시지."""

    message: str
    timestamp: str


class MessageSend(WebSocketEnvelope[MessageSendPayload]):
    type: Literal["MESSAGE_SEND"]

    def __init__(self, message: str) -> None:
        super().__init__(
            type=WebSocketMessageTypes.MESSAGE_SEND,
            payload=MessageSendPayload(message=message),
        )


class MessageReceive(WebSocketEnvelope[MessageReceivePayload]):
    type: Literal["MESSAGE_RECEIVE"]

    def __init__(self, message: str, nickname: str, timestamp: str) -> None:
        super().__init__(
            type=WebSocketMessageTypes.MESSAGE_RECEIVE,
            payload=MessageReceivePayload(message=message, nickname=nickname, timestamp=timestamp),
        )


class ServerStatusRequest(WebSocketEnvelope[ServerStatusRequestPayload]):
    type: Literal["SERVERSTATUS_REQUEST"]

    def __init__(self) -> None:
        super().__init__(
            type=WebSocketMessageTypes.SERVER_STATUS_REQUEST,
            payload=ServerStatusRequestPayload(),
        )


class ServerStatusResponse(WebSocketEnvelope[ServerStatusResponsePayload]):
    type: Literal["SERVERSTATUS_RESPONSE"]

    def __init__(self, client_count: int) -> None:
        super().__init__(
            type=WebSocketMessageTypes.SERVER_STATUS_RESPONSE,
            payload=ServerStatusResponsePayload(clientCount=client_count),
        )


class SystemMessageReceive(WebSocketEnvelope[SystemMessageReceivePayload]):
    type: Literal["SYSTEM_MESSAGE_RECEIVE"]

    def __init__(self, message: str, timestamp: str) -> None:
        super().__init__(
            type=WebSocketMessageTypes.SYSTEM_MESSAGE_RECEIVE,
            payload=SystemMessageReceivePayload(message=message, timestamp=timestamp),
        )

File: chat-client/src/messages.ts
export const WebSocketMessageTypes = {
  MESSAGE_SEND: "MESSAGE_SEND",
  MESSAGE_RECEIVE: "MESSAGE_RECEIVE",
  SERVER_STATUS_REQUEST: "SERVERSTATUS_REQUEST",
  SERVER_STATUS_RESPONSE: "SERVERSTATUS_RESPONSE",
  SYSTEM_MESSAGE_RECEIVE: "SYSTEM_MESSAGE_RECEIVE",
} as const;

export function buildMessageSend(message: string): unknown {
  return {
    type: WebSocketMessageTypes.MESSAGE_SEND,
    payload: { message },
  };
}

export function buildServerStatusRequest(): unknown {
  return {
    type: WebSocketMessageTypes.SERVER_STATUS_REQUEST,
    payload: {},
  };
}

export function formatInboundMessage(payload: unknown): string {
  if (typeof payload !== "object" || payload === null) {
    return String(payload);
  }

  const type = String((payload as Record<string, unknown>).type ?? "").toUpperCase();
  const data = (payload as Record<string, unknown>).payload;
  if (type === WebSocketMessageTypes.MESSAGE_RECEIVE && isPlainObject(data)) {
    const { nickname, message, timestamp } = data;
    if (typeof message === "string") {
      const author = typeof nickname === "string" ? nickname : "unknown";
      const ts = typeof timestamp === "string" ? timestamp : "";
      return `${author}${ts ? `@${ts}` : ""}: ${message}`;
    }
  }

  if (type === WebSocketMessageTypes.SERVER_STATUS_RESPONSE && isPlainObject(data)) {
    const count = data.clientCount;
    if (typeof count === "number") {
      return `[서버 인원] 현재 접속자 ${count}명`;
    }
  }

  if (type === WebSocketMessageTypes.SYSTEM_MESSAGE_RECEIVE && isPlainObject(data)) {
    const { message, timestamp } = data;
    if (typeof message === "string") {
      const ts = typeof timestamp === "string" ? timestamp : "";
      return `[system${ts ? `@${ts}` : ""}] ${message}`;
    }
  }

  return JSON.stringify(payload);
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}
